plugins/jenkins-backend/src/service/router.ts

```diff
@@ -0,0 +1,124 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import { errorHandler } from '@backstage/backend-common';
+import express from 'express';
+import Router from 'express-promise-router';
+import { Logger } from 'winston';
+import { JenkinsInfoProvider } from './jenkinsInfoProvider';
+import { JenkinsApiImpl } from './jenkinsApi';
+
+export interface RouterOptions {
+  logger: Logger;
+  jenkinsInfoProvider: JenkinsInfoProvider;
+}
+
+export async function createRouter(
+  options: RouterOptions,
+): Promise<express.Router> {
+  const { jenkinsInfoProvider } = options;
+
+  const jenkinsApi = new JenkinsApiImpl();
+
+  const router = Router();
+  router.use(express.json());
+
+  router.get(
+    '/v1/entity/:namespace/:kind/:name/projects',
+    async (request, response) => {
+      const { namespace, kind, name } = request.params;
+      const branch = request.query.branch;
+      let branchStr: string | undefined;
+
+      if (branch === undefined) {
+        branchStr = undefined;
+      } else if (typeof branch === 'string') {
+        branchStr = branch;
+      } else {
+        // this was passed in as something weird -> 400
+        // https://evanhahn.com/gotchas-with-express-query-parsing-and-how-to-avoid-them/
+        response
+          .status(400)
+          .send('Something was unexpected about the branch queryString');
+
+        return;
+      }
+
+      const jenkinsInfo = await jenkinsInfoProvider.getInstance({
+        entityRef: {
+          kind,
+          namespace,
+          name,
+        },
+      });
+      const projects = await jenkinsApi.getProjects(jenkinsInfo, branchStr);
+
+      response.json({
+        projects: projects,
+      });
+    },
+  );
+
+  router.get(
+    '/v1/entity/:namespace/:kind/:name/job/:jobName/:buildNumber',
+    async (request, response) => {
+      const { namespace, kind, name, jobName, buildNumber } = request.params;
+
+      const jenkinsInfo = await jenkinsInfoProvider.getInstance({
+        entityRef: {
+          kind,
+          namespace,
+          name,
+        },
+        jobName,
+      });
+
+      const build = await jenkinsApi.getBuild(
+        jenkinsInfo,
+        jobName,
+        parseInt(buildNumber, 10),
+      );
+
+      response.json({
+        build: build,
+      });
+    },
+  );
+
+  router.post(
+    '/v1/entity/:namespace/:kind/:name/job/:jobName/:buildNumber::rebuild',
+    async (request, response) => {
+      const { namespace, kind, name, jobName } = request.params;
+
+      const jenkinsInfo = await jenkinsInfoProvider.getInstance({
+        entityRef: {
+          kind,
+          namespace,
+          name,
+        },
+        jobName,
+      });
+
+      await jenkinsApi.buildProject(jenkinsInfo, jobName);
+
+      // TODO: return the buildNumber which was started.
+      response.json({});
+    },
+  );
+
+  router.use(errorHandler());
```

Comment:
as a kind of a side note, we actually just had an interesting discussion among ourselves about whether the error handler should belong inside the plugin, or outside in the backend where the adopter can control it (and give it custom parameters etc, or use an alternate one). We are leaning toward putting it outside FYI. You don't have to change this right now tho, we may have a go-over and changing this consistently in the repo some day later